scipy/signal/wavelets.py

```diff
@@ -352,14 +441,22 @@ def cwt(data, wavelet, widths):
     >>> sig  = np.cos(2 * np.pi * 7 * t) + signal.gausspulse(t - 0.4, fc=2)
     >>> widths = np.arange(1, 31)
     >>> cwtmatr = signal.cwt(sig, signal.ricker, widths)
-    >>> plt.imshow(cwtmatr, extent=[-1, 1, 31, 1], cmap='PRGn', aspect='auto',
+    >>> plt.imshow(cwtmatr, extent=[-1, 1, 1, 31], cmap='PRGn', aspect='auto',
     ...            vmax=abs(cwtmatr).max(), vmin=-abs(cwtmatr).max())
     >>> plt.show()
-
     """
-    output = np.zeros([len(widths), len(data)])
+    # Determine output type
+    if dtype is None:
+        wavelet_type = np.asarray(wavelet(1, widths[0], **kwargs)).dtype.char
+        data_type = np.asarray(data).dtype.char
+        if wavelet_type in 'FD' or data_type in 'FD':
+            dtype = np.complex128
+        else:
+            dtype = np.float64
+
+    output = np.zeros((len(widths), len(data)), dtype=dtype)
     for ind, width in enumerate(widths):
-        wavelet_data = wavelet(min(10 * width, len(data)), width)
-        output[ind, :] = convolve(data, wavelet_data,
-                                  mode='same')
+        N = np.min([10 * width, len(data)])
+        wavelet_data = np.conj(wavelet(N, width, **kwargs)[::-1])
```

Comment:
I take it that this time-reversal change is a bug we've had around for a long time. Do you have a good reference for the time-reversal+conjugate formulation that you can put in the `Notes` section of `cwt`?

Reply:
The discrete wavelet transform is described in depth in S. Mallat's *A Wavelet Tour of Signal Processing*. I'll add a reference.

Reply:
This sounds convincing to me. Our algorithm has a proper reference now. Does this guarantee that there is no backwards-incompatible change? I guess that's the same as asking if we have to consider asymmetric wavelet functions if the symmetric ones are unchanged by the adjustment.

In the absence of clarification, I'll probably add a compatibility-related note to the release notes.

Reply:
The use of time-reversal and conjugation is consistent with what is in PyWavelets 1.1.0 (older versions of PyWavelets were missing the conjugate).
The actual numerical implementation in PyWavelets is a bit different from `cwt` as defined here. This one is much easier to follow. You can see that the result is consistent for SciPy vs. PyWavelets based on the images shown in this issue:
PyWavelets/pywt#531
which has links related to the numerical difference in the algorithm.